core/services/pipeline/keypath.go

```diff
@@ -0,0 +1,47 @@
+package pipeline
+
+import (
+	"bytes"
+
+	"github.com/pkg/errors"
+)
+
+var (
+	ErrWrongKeypath  = errors.New("wrong keypath format")
+	KeypathSeparator = []byte(".")
+)
+
+// Keypath contains keypath parsed by NewKeypathFromString.
+type Keypath struct {
+	NumParts int    // can be 0, 1 or 2
+	Part0    string // can be empty string if NumParts is 0
+	Part1    string // can be empty string if NumParts is 0 or 1
+}
+
+// NewKeypathFromString creates a new Keypath from the given string.
+// Returns error if it fails to parse the given keypath string.
+func NewKeypathFromString(keypathStr string) (Keypath, error) {
```

Comment:
All the pipeline code relies on at most two elements. Even the prior solution used `[2][]byte`.
If you want to change this to support "arbitrary depth", then Okay but this would deserve a dedicated feature, which I can do but not sooner than I finish all pipeline refactoring.

Reply:
Also I wanted to clarify that the goal of this work was to refactor the code, add tests and do not change semantic.